Incremental commits spec changes (#1511)
input-output-hk/hydra: spec/offchain.tex

	\item $\hats$: Sequence number of latest seen snapshot.
	\item $\hatSigma \in {(\tyNatural \times \tyBytes)}^{*}$:
		Accumulator of signatures of the latest seen snapshot, indexed by parties.
	\item $\hatmL$: UTxO set representing the local ledger state resulting from
		applying $\hatmT$ to $\bar{S}.U$ to validate requested transactions.
	\item $\hatmT \in \mT^{*}$: List of transactions applied locally and pending
		inclusion in a snapshot (if this party is the next leader).
	\item \red{$U_\alpha \in \gcUTXOset$: Pending increment UTxO to be deposited to the head. Can be bottom if nothing is pending to be deposited.}
	\item $U_\omega \in \gcUTXOset$: Pending decrement UTxO to be withdrawn from the head. Can be bottom if nothing is pending to be withdrawn.
	\item $\bar{\mc S}$: Snapshot object of the latest confirmed snapshot which contains:
		\begin{center}
			\begin{tabular}{|l|l|}\hline
				$\bar{\mc S}.v$         & snapshot version \\ \hline
				$\bar{\mc S}.s$         & snapshot number \\ \hline
				$\bar{\mc S}.T$         & set of transactions relating this snapshot to its predecessor \\ \hline
				$\bar{\mc S}.U$         & snapshotted UTxO set \\ \hline

> 				\red{$\bar{\mc S}.U_\alpha$}  & \red{peding UTxO to commit} \\ \hline
> 				\red{$\bar{\mc S}.U_\omega$}  & \red{peding UTxO to decommit} \\ \hline

Reviewer: ```suggestion
				\red{$\bar{\mc S}.U_\alpha$}  & \red{pending UTxO to commit} \\ \hline
				\red{$\bar{\mc S}.U_\omega$}  & \red{pending UTxO to decommit} \\ \hline
```

				$\bar{\mc S}.\sigma$    & multisignature \\ \hline
			\end{tabular}
		\end{center}

> 		The function $\Sno(v,n,T,U,\red{U_\alpha}, U_\omega)$ initializes a snapshot object with no multi-signature set.

Reviewer: Must: fix inconsistency in argument ordering and what to pass here and in all later mentions of this.

Here it is given Utxos, in the off-chain figure (on ack sn) it is given transactions, and in the description of handling Ack Sn it is given utxo hashes $\eta$.

\begin{itemize}
	\item $\hatv$: Last seen open state version.
	\item $\hats$: Sequence number of latest seen snapshot.
	\item $\hatSigma \in {(\tyNatural \times \tyBytes)}^{*}$:
		Accumulator of signatures of the latest seen snapshot, indexed by parties.
	\item $\hatmL$: UTxO set representing the local ledger state resulting from
		applying $\hatmT$ to $\bar{S}.U$ to validate requested transactions.
	\item $\hatmT \in \mT^{*}$: List of transactions applied locally and pending
		inclusion in a snapshot (if this party is the next leader).
	\item \red{$U_\alpha \in \gcUTXOset$: Pending increment UTxO to be deposited to the head.}
	\item $U_\omega \in \gcUTXOset$: Pending decrement UTxO to be withdrawn from the head.
	\item \red{$U_\theta \in U_\alpha \cup U_\omega \cup \bot$: Either a decrement UTxO, an increment UTxO or $\bot$.
		May be $\bot$ if nothing is pending to be applied.}

> 	\item \red{$\theta \in Normal \cup Inc \cup Dec$: A type of a snapshot object.}

Reviewer: Instead of defining $U_\alpha$ and $U_\omega$, you should unify the types directly (otherwise: what's the point of having a unified $U_\theta$?) 

Maybe using something like:

$U_\theta \in \mathsf{Normal} \cup (\mathsf{Inc}, \mathsf{UTxOs}) \cup (\mathsf{Dec}, \mathsf{UTxOs})$

(We should really define $\mathsf{UTxOs}$ better)

Reviewer: Also, I think you could avoid the dedicated $\theta$ and just define that $\eta_\theta$ includes that discriminator?

	\item \red{$\bar{\mc S}\theta \in \theta$: A type of a snapshot object $\bar{\mc S}$.
			Snapshot is of Normal type if it doesn't contain anything to increment or decrement and
		Inc/Dec types signal there is something to increment or decrement respectfully.
			}
	\item $\bar{\mc S}$: Snapshot object of the latest confirmed snapshot which contains:
		\begin{center}
			\begin{tabular}{|l|l|}\hline
				$\bar{\mc S}.v$         & snapshot version \\ \hline
				$\bar{\mc S}.s$         & snapshot number \\ \hline
				$\bar{\mc S}.T$         & set of transactions relating this snapshot to its predecessor \\ \hline
				$\bar{\mc S}.U$         & snapshotted UTxO set \\ \hline

> 				\red{$\bar{\mc S}.U_\theta$}  & \red{peding increment/decrement UTxO} \\ \hline

Reviewer: ```suggestion
				\red{$\bar{\mc S}.U_\theta$}  & \red{pending increment/decrement UTxO} \\ \hline
```

	      applying $\hatmT$ to $\bar{S}.U$ to validate requested transactions.
	\item $\hatmT \in \mT^{*}$: List of transactions applied locally and pending
		inclusion in a snapshot (if this party is the next leader).
	\item \red{$\tx_\alpha \in \mathcal{T}$: Pending increment transaction, whose inputs are to be deposited to the head.}
	\item $\tx_\omega \in \mathcal{T}$: Pending decrement transaction, whose outputs are to be withdrawn from the head.
	\item $\bar{\mc S}$: Snapshot object of the latest confirmed snapshot which contains:
		\begin{center}
			\begin{tabular}{|l|l|}\hline
				$\bar{\mc S}.v$         & snapshot version \\ \hline
				$\bar{\mc S}.s$         & snapshot number \\ \hline
				$\bar{\mc S}.T$         & set of transactions relating this snapshot to its predecessor \\ \hline
				$\bar{\mc S}.U$         & snapshotted UTxO set \\ \hline
				\red{$\bar{\mc S}.U_\alpha = \mathsf{inputs}(\tx_\alpha)$}  & \red{pending UTxO to commit} \\ \hline
				\red{$\bar{\mc S}.U_\omega = \mathsf{outputs}(\tx_\omega)$}  & \red{pending UTxO to decommit} \\ \hline
				$\bar{\mc S}.\sigma$    & multisignature \\ \hline
			\end{tabular}
		\end{center}

>     The function $\Sno(v,n,U,T,\red{\mathsf{inputs}(tx_\alpha)}, \mathsf{outputs}(tx_\omega))$ initializes a snapshot object with no multi-signature set.

Reviewer: Should: move the $\mathsf{inputs}(tx_\alpha)$ and $\mathsf{outputs}(tx_\omega)$ to the invocation

transaction, the parties compute $\Uinit \gets \bigcup_{j=1}^{n} C_j$ using previously
observed $C_j$ and initialize $\hatmL = \Uinit$. The seen transaction set is
initialized empty $\hatmT = \emptyset$, seen head state version $\hatv = 0$, as
well as snapshot number $\hats = 0$. No \red{increment/}decrement UTxO is pending

> ($U_\alpha = \bot \land  U_\omega = \bot$) and the initial snapshot object is defined accordingly

Reviewer: Are we keeping now UTxOs or transactions in our state?

Here we mention $U_\alpha$, in the figure it is $tx_{\omega}$ and the "Variables" section explains both.

Must: keep only transaction **or** UTxO

receiving party $\party_{i}$ is the next snapshot
leader, a message to request snapshot signatures $\hpRS$ is sent. \\

> \dparagraph{$\hpRD$.}\quad Upon receiving request $(\hpRD,\tx_\omega)$, \red{if
>   another decrement (or increment) is pending to be settled the transaction is
>   rejected. Otherwise} the transaction is checked against the \emph{local}
>   ledger state and if it is not applicable yet or another decommit is still, the protocol does $\KwWait$ to
>   retry later or eventually marks the decommit as invalid. \red{After applying $\tx$, it's inputs

Reviewer: Should: fix this sentence. It was already mentioning pending decommits. Let's just extend the sentence there.

```suggestion
\dparagraph{$\hpRD$.}\quad Upon receiving request $(\hpRD,\tx_\omega)$, the transaction is checked against the \emph{local}
  ledger state and if it is not applicable yet or another commit or decommit is pending still, the protocol does $\KwWait$ to
  retry later or eventually marks the decommit as invalid. \red{After applying $\tx$, it's inputs
```

  are removed from \emph{local} ledger state so that they are not available any
  more and the outputs are kept in the local state ($U_\omega$) to be posted
  on-chain in the next snapshot.} Finally, if there is no current snapshot in flight
  ($\hats = \bar{\mc S}.s$) and the receiving party $\party_{i}$ is the next
  snapshot leader, a message to request snapshot signatures $\hpRS$ containing
  the decrement transaction $\tx_\omega$ is sent. \\

\dparagraph{$\mathtt{decrementTx}$.}\quad Upon observing the \mtxDecrement{}
transaction, which removed outputs $U$ from the head, the corresponding
pending decrement transaction is cleared and the observed version $v$ is used for future
snapshots by setting $\hatv \gets v$.  Note that the version of the open head state
is incremented on each \mtxDecrement{} transaction as described in Section~\ref{sec:decrement-tx}
\todo{missing to mention the multicast of $\hpRS$} \\

\red{\dparagraph{$\hpRI$.}\quad Upon receiving request $(\hpRI,\tx)$, if
  another decrement (or increment) is pending to be settled the transaction is

>   rejected. Otherwise the transaction inputs are kept in the local state ($U_\alpha$) to be included

Reviewer: It is not strictly "rejected", but we wait for it to not be the case. Let's use wording similar to the `ReqTx` and `RecDec`:

"If another commit or decommit is pending, the protocol does wait to retry later or eventually marks the decommit as invalid."

  transaction is cleared and the observed version is used for future snapshots by setting
  $\hatv = v$. The version of the open head state is incremented on each \mtxIncrement{}
  transaction as described in Section~\ref{sec:increment-tx}
}
\todo{missing to mention the multicast of $\hpRS$} \\

\dparagraph{$\hpRS$.}\quad Upon receiving request
$(\hpRS,v,s,\mT_{\mathsf{req}}, \red{\tx_\alpha}, \tx_\omega)$\footnote{Snapshot requests
	with only transaction identifiers are possible too if all parties keep an
	index of previously seen transactions and their identifiers.} from party
$\party_j$, the receiving $\party_i$ $\Req$s that $v$ refers to the current
open state version, $s$ is the next snapshot number
and that party $\party_j$ is responsible for leading its creation.\todo{define
	$\hpLdr$} Party $\party_i$ may has to $\KwWait$ until the previous snapshot is
  confirmed ($\bar{\mc S}.s = \hats$). Then either:
\begin{itemize}
  \item
    there is a pending decrement and we check that $\tx_\omega$ transaction is applicable to the last confirmed
	  UTxO set $\bar{\mc S}.U$ and it's outputs are removed, yielding the still active UTxO set $U_{\mathsf{active}}$
\red{\item

>     there is a pending increment $\tx_\alpha$ which inputs will be made available in the active UTxO set

Reviewer: Should: not add this here (and remove the itemization again). There is nothing done on `reqSn` related to the increment.

  ledger state and if it is not applicable yet, the protocol does $\KwWait$ to
  retry later or output an error message.
  If it is applicable, then the transaction outputs are kept in the local state ($U_\theta$)
  to be posted on-chain in the next snapshot.
  Finally, if there is no current snapshot in flight ($\hats = \bar{\mc S}.s$) and the
  receiving party $\party_{i}$ is the next snapshot leader, a message to request
  snapshot signatures $\hpRS$ containing the increment transaction $\tx_\alpha$ is sent.} \\

\red{\dparagraph{$\mathtt{incrementTx}$.}\quad Upon observing the \mtxIncrement{}
	transaction which outputs are added to the confirmed UTxO in the Head, we include
  the outputs to the \emph{local} ledger state so that they become available.
  Finally the corresponding pending increment
  transaction is cleared and the observed version is used for future snapshots by setting
  $\hatv = v$. The version of the open head state is incremented on each \mtxIncrement{}
  transaction as described in Section~\ref{sec:increment-tx}
}
\todo{missing to mention the multicast of $\hpRS$} \\

\dparagraph{$\hpRS$.}\quad Upon receiving request

> $(\hpRS,v,s,\mT_{\mathsf{req}}, \red{\psi})$\footnote{Snapshot requests

Reviewer: What is $\psi$?
